crm-project/src/main/java/CRM/service/UserService.java

```diff
@@ -149,28 +151,54 @@ public List<User> getAllInBoard(long boardId) throws AccountNotFoundException {
      * @throws IllegalArgumentException if the combination of the given user and board already exists in the database
      */
-    public void addUserToBoard(long userId, long boardId) throws AccountNotFoundException {
+    public List<User> updateUserToBoard(long userId, long boardId, long permissionId) throws AccountNotFoundException {
         User user;
         Board board;
         try {
             user = Validations.doesIdExists(userId, userRepository);
             board = Validations.doesIdExists(boardId, boardRepository);
-
         } catch (NoSuchElementException e) {
             throw new AccountNotFoundException(ExceptionMessage.ACCOUNT_DOES_NOT_EXISTS.toString());
         }
 
-        NotificationSetting notificationSetting = Validations.doesIdExists(2L, settingRepository);
+        if(user.equals(board.getCreatorUser())){
+            throw new IllegalArgumentException(ExceptionMessage.ADMIN_CANT_CHANGE_HIS_PERMISSION.toString());
+        }
 
-        UserPermission userPermission = new UserPermission();
-        userPermission.setId(0L);
-        userPermission.setUser(user);
-        userPermission.setPermission(Permission.USER);
+        UserPermission userPermissionInBoard = null;
+        Set<UserPermission> userPermissionsSet = board.getUsersPermissions();
 
-        board.addUserPermissionToBoard(userPermission);
+        for (UserPermission userInBoard : userPermissionsSet) {
+            if (userInBoard.getId().equals(userId)) {
+                userPermissionInBoard = userInBoard;
+                break;
+            }
+        }
 
-        createDefaultSettingForNewUserInBoard(user, board, notificationSetting);
+        Permission permissionRequest = Permission.values()[(int) permissionId];
+
+        if (userPermissionInBoard == null) { //create new
+            if(!permissionRequest.equals(Permission.ADMIN) && !permissionRequest.equals(Permission.UNAUTHORIZED)) createNewUserPermission(user, permissionRequest, board);
+            else throw new IllegalArgumentException(ExceptionMessage.CANT_ASSIGN_PERMISSION.toString());
+        } else if (permissionRequest.equals(Permission.ADMIN)) {
+            throw new IllegalArgumentException(ExceptionMessage.PERMISSION_NOT_ALLOWED.toString());
+        } else if (permissionRequest.equals(Permission.USER) && !userPermissionInBoard.getPermission().equals(Permission.USER)) {
+            userPermissionInBoard.setPermission(permissionRequest);
+        } else if (permissionRequest.equals(Permission.LEADER) && !userPermissionInBoard.getPermission().equals(Permission.LEADER)) {
+            userPermissionInBoard.setPermission(permissionRequest);
+        } else if (permissionRequest.equals(Permission.UNAUTHORIZED)) {
+            userPermissionsSet.remove(userPermissionInBoard);
+        } else {
+            throw new IllegalArgumentException(ExceptionMessage.USER_ALREADY_HAS_THIS_PERMISSION.toString());
+        }
```

Comment:
Same here put in an outside private function.

Also, start from the beginning:

// YOU WONT ACCEPT THIS:
if(permission.equals(admin)) -> throw exception

// YOU WILL ACCEPT THIS:
delete: if(permission.equals(unauthorized)) -> remove from list (make sure he exists before remove)
create: if(user == null) -> create a new user and push to the db with "user" permission regardless of the request permission

after you handled all the edge cases (remove / add / not admin), you can get to the update without validating the permission because you know the following: 
- the user is not null -> means he exists in the database
- the user is not going to be an admin because you'd throw an exception
- the user is authorized because otherwise you'd remove him from the db before here

userPermission.setPermission(permission)